### ISSUE TITLE
More sample scripts and the third has piping which may work now?

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#8533]
@lloyddewit I am taking advantage of your last week, to raise a point I don't see yet in your list of script tasks!

The [site is here](https://waterdata.usgs.gov/blog/beyond-basic-plotting/): 

<img width="773" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/554c593a-bc95-4d14-855b-30c33cf252c1">

It is useful as being climatic.  And it is a sort of training exercise that uses ggplot in the ordinary way.  So useful to see if the code can run.  It startsd with code that needs a new package to be loaded.  I did that.

Here is the code:

```
library(dataRetrieval)
library(dplyr) # for `rename` & `select`
library(tidyr) # for `gather`
library(ggplot2)

# Get the data by giving site numbers and parameter codes
# 00060 = stream flow, 00530 = total suspended solids, 00631 = concentration of inorganic nitrogen
wi_daily_wq <- readNWISdv(siteNumbers = c("05430175", "05427880", "05427927"),
                          parameterCd = c("00060", "00530", "00631"),
                          startDate = "2017-08-01", endDate = "2017-08-31")

# Clean up data to have human-readable names + move data into long format
wi_daily_wq <- renameNWISColumns(wi_daily_wq,
                                 p00530 = "TSS",
                                 p00631 = "InorganicN") %>%
  select(-ends_with("_cd")) %>%
  gather(key = "parameter", value = "value", -site_no, -Date)

# Setup plot without facets
p <- ggplot(data = wi_daily_wq, aes(x = Date, y = value)) +
  geom_point(aes(color = site_no)) +
  theme_bw()

# Now, we can look at the plot and see how it looks before we facet
# Obviously, the scales are off because we are plotting flow with concentrations
p

# Add vertical facets, aka divide the plot up vertically since they share an x axis
p + facet_grid(parameter ~ .)

# Add vertical facets, but scale only the y axes freely
p + facet_grid(parameter ~ ., scales = "free_y")

p + facet_grid(parameter ~ ., scales = "free_y",
               switch = "y", # flip the facet labels along the y axis from the right side to the left
               labeller = as_labeller( # redefine the text that shows up for the facets
                 c(Flow = "Flow, cfs", InorganicN = "Inorganic N, mg/L", TSS = "TSS, mg/L"))) +
  ylab(NULL) + # remove the word "values"
  theme(strip.background = element_blank(), # remove the background
        strip.placement = "outside") # put labels to the left of the axis text
        
```
It runs.  It includes lots of comments.  They are like the comments we include in the dialogs.  But they do not appear in the output window.  Please could they?  Note that some comments are at the end of command lines.  If commands are on, then those commands are put into the output window.  

I assume it may be difficult to put them into the output window, without the commands?  That would be a bit disappointing, but I would be happy with that for now.

Then, on the extremely positive side, it all runs fine.   As mentioned recently it puts the (graphical) results into the R viewer.
It is also useful, because it mis retrieving hydrological data from the USGS (US Geological Service).  If people can use scripts then this might be all we provide initially, i.e. some examples.

The next section is essentially tidying the data - neat code, and perhaps we will sometimes continue to have that in code.
Then, just before the line `# Setup plot without facets` we can read the data into an R-Instat data book, and that works fine.

Then I hope we can do the same graphs from our dialogs.  If not, then we might, for now, have to "top" and "tail" each of the graphs with the code needed to get a data frame, and then to add the plot into our "system".

I see work already specified here.  The addition of comments to the output window is the only @lloyddewit query - perhaps to implement, or at least to discuss and add to the list of remaining tasks.

**Comment by lloyddewit** [comment]
@rdstern Thank you for reporting.
Including code comments in the output window will be fixed in PR #8494. So I don't think we need to add this issue to our summary of script window changes (issue #7989).

**Comment by rdstern** [comment]
@lloyddewit thanks - looking forward

**Comment by rdstern** [comment]
@lilyclements and @lloyddewit here is another sample script and perhaps another point on the output window - from the script window, to consider.  But also this is quite a complicated script of the sort we can be using and the handling seems to be working well.  

@lilyclements I am not sure if this is an interesting package for us for the display of tables.  One of the vignettes on gt mentioned extra packages.  By the way gt has been updated (to 0.9) since the new master was produced.  I am not sure what is new, but it has made more data sets available.

The extra packages are grExtras (already included), gtsummary we know about, plus pointblank, tfrmt and gto.  gto is deveoted to putting gt output into word documents.  tfrmt seemed possibly interesting?  It is about producing the right tables before you have the data.  Then - I assume - as soon as you have the data - then hey-presto you have the table.

So I tried an example from [here](https://gsk-biostatistics.github.io/tfrmt/articles/examples.html#efficacy).  Note it is a team from Glaxo Smith Klein - and I really like the idea of sorting out the format first and then being able to quickly produce an excellent report!

These commands below are quite complicated - so I also wanted to see if the code runs!

```
library(tfrmt)
library(dplyr)
library(gt)
library(tidyr)
tfrmt(
  # specify columns in the data
  group = c(rowlbl1,grp),
  label = rowlbl2,
  column = column, 
  param = param,
  value = value,
  sorting_cols = c(ord1, ord2)) %>% 
  print_to_gt(data_demog) %>% 
  tab_options(
    container.width = 900
  )
  
  ##run to here for the first table - really messy
  
  tfrmt(
  # specify columns in the data
  group = c(rowlbl1,grp),
  label = rowlbl2,
  column = column, 
  param = param,
  value = value,
  sorting_cols = c(ord1, ord2),
  # specify value formatting 
  body_plan = body_plan(
    frmt_structure(group_val = ".default", label_val = ".default", frmt_combine("{n} ({pct} %)", 
                                                                                n = frmt("xxx"),
                                                                                pct = frmt("xx.x"))),
    frmt_structure(group_val = ".default", label_val = "n", frmt("xxx")),
    frmt_structure(group_val = ".default", label_val = c("Mean", "Median", "Min","Max"), frmt("xxx.x")),
    frmt_structure(group_val = ".default", label_val = "SD", frmt("xxx.xx")),
    frmt_structure(group_val = ".default", label_val = ".default", p = frmt_when(">0.99" ~ ">0.99",
                                                                                 "<0.001" ~ "<0.001",
                                                                                 TRUE ~ frmt("x.xxx", missing = "")))
  )) %>% 
  print_to_gt(data_demog) %>% 
  tab_options(
    container.width = 900
  )
 ##run to here for the second table
  
  
  tfrmt(
  # specify columns in the data
  group = c(rowlbl1,grp),
  label = rowlbl2,
  column = column, 
  param = param,
  value = value,
  sorting_cols = c(ord1, ord2),
  # specify value formatting 
  body_plan = body_plan(
    frmt_structure(group_val = ".default", label_val = ".default", frmt_combine("{n} {pct}", 
                                                                                n = frmt("xxx"),
                                                                                pct = frmt_when("==100" ~ "",
                                                                                                "==0" ~ "",
                                                                                                TRUE ~ frmt("(xx.x %)")))),
    frmt_structure(group_val = ".default", label_val = "n", frmt("xxx")),
    frmt_structure(group_val = ".default", label_val = c("Mean", "Median", "Min","Max"), frmt("xxx.x")),
    frmt_structure(group_val = ".default", label_val = "SD", frmt("xxx.xx")),
    frmt_structure(group_val = ".default", label_val = ".default", p = frmt("")),
    frmt_structure(group_val = ".default", label_val = c("n","<65 yrs","<12 months","<25"), p = frmt_when(">0.99" ~ ">0.99",
                                                                                 "<0.001" ~ "<0.001",
                                                                                 TRUE ~ frmt("x.xxx", missing = "")))
  ),
  # remove extra cols
  col_plan = col_plan(-grp, 
                      -starts_with("ord") )) %>% 
  print_to_gt(data_demog) %>% 
  tab_options(
    container.width = 900
  )
  ##run to here for the third table
  
  
  data_demog %>% 
  distinct(rowlbl1,grp)
  
  ## this appears in the output window
  
  
  tfrmt(
  # specify columns in the data
  group = c(rowlbl1,grp),
  label = rowlbl2,
  column = column, 
  param = param,
  value = value,
  sorting_cols = c(ord1, ord2),
  # specify value formatting 
  body_plan = body_plan(
    frmt_structure(group_val = ".default", label_val = ".default", frmt_combine("{n} {pct}", 
                                                                                n = frmt("xxx"),
                                                                                pct = frmt_when("==100" ~ "",
                                                                                                "==0" ~ "",
                                                                                                TRUE ~ frmt("(xx.x %)")))),
    frmt_structure(group_val = ".default", label_val = "n", frmt("xxx")),
    frmt_structure(group_val = ".default", label_val = c("Mean", "Median", "Min","Max"), frmt("xxx.x")),
    frmt_structure(group_val = ".default", label_val = "SD", frmt("xxx.xx")),
    frmt_structure(group_val = ".default", label_val = ".default", p = frmt("")),
    frmt_structure(group_val = ".default", label_val = c("n","<65 yrs","<12 months","<25"), p = frmt_when(">0.99" ~ ">0.99",
                                                                                 "<0.001" ~ "<0.001",
                                                                                 TRUE ~ frmt("x.xxx", missing = "")))
  ),
  # remove extra cols
  col_plan = col_plan(-grp, 
                      -starts_with("ord") ),
  # Specify column styling plan
  col_style_plan = col_style_plan(
    col_style_structure(align = c(".",","," "), col = vars(everything()))
  ),
  
    # Specify row group plan
  row_grp_plan = row_grp_plan(
    row_grp_structure(group_val = ".default", element_block(post_space = " ")),
    label_loc = element_row_grp_loc(location = "column")
  )
  
  ) %>% 
  print_to_gt(data_demog) %>% 
  tab_options(
    container.width = 900
  )
  
  ##run to here for the final table - still looks a bit messy to me, but same as the article.
```

The only issue I could find is that the results appear in a browser, rather than in our output window.  I haven't checked the command options at all, so just note this "feature" at this stage.

**Comment by rdstern** [comment]
@lloyddewit  I am using code from [here](https://gt.rstudio.com/articles/intro-creating-gt-tables.html).

@lilyclements I am keen for us to be able to at least produce gt tables from the scripts they provide.  Here is a picture and it imports fine into R-Instat as I also show:

<img width="592" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/2739225b-dee3-4665-959e-19e5b53953c1">

In R-Instat it imports and magically the piping and other operators are transformed ok.  This is the code in R-Instat.

```
# Take the `islands` dataset and use some
# dplyr functionality to obtain the ten
# biggest islands in the world
islands_tbl <- 
  tibble(
    name = names(islands),
    size = islands
  ) |>
  arrange(desc(size)) |>
  slice(1:10)

# Display the table
islands_tbl
#> # A tibble: 10 × 2
#>    name           size
#>    <chr>         <dbl>
#>  1 Asia          16988
#>  2 Africa        11506
```

In this code, from R version 4.1 I understand that |> is now the new piping operator and added to base R.  It used to be from a package and given as %>%.  

@lloyddewit I don't think we yet process this new operator.  So this code gives errors.  

So I then tried with the original piping operator:

```
library(gt)
library(tidyverse)
# Take the `islands` dataset and use some
# dplyr functionality to obtain the ten
# biggest islands in the world
islands_tbl <- 
  tibble(
    name = names(islands),
    size = islands
  ) %>%
  arrange(desc(size)) %>%
  slice(1:10)

# Display the table
islands_tbl
    
   # data_book$import_data(data_tables=list(islands_tbl=islands_tbl))
 
   # Create a display table showing ten of
# the largest islands in the world
gt_tbl <- gt(islands_tbl)

# Show the gt Table
gt_tbl
```

Now it does not give the errors.  But it also does not recognise the table it should have created.  So the lines:

```
# Display the table
islands_tbl 

```

give an error that it doesn't have the table?

Aha, but this version does work, i.e. when I put the old piping operator and the whole thing on one line.  @lloyddewit I am very relieved, and maybe you already told me that this is still part of the lines going to statements work that is still to be completed? 

```
library(gt)
library(tidyverse)
# Take the `islands` dataset and use some
# dplyr functionality to obtain the ten
# biggest islands in the world
islands_tbl <- tibble(name = names(islands),size = islands ) %>% arrange(desc(size)) %>%  slice(1:10)

# Display the table
islands_tbl
    
   # data_book$import_data(data_tables=list(islands_tbl=islands_tbl))
 
   # Create a display table showing ten of
# the largest islands in the world
gt_tbl <- gt(islands_tbl)

# Show the gt Table
gt_tbl
```
 This is still all very exciting indeed!

**Comment by lloyddewit** [comment]
@rdstern I will add the new pipe operator to PR #8494.
The multiline issue above should also be fixed in this PR. I will test.

**Comment by lloyddewit** [comment]
The second big block of code above is not correctly parsed by RScript. I raised an issue for this: lloyddewit/RScript#16.
If I manually correct the statement, then there is a second problem: the output is not correctly displayed. I will investigate this when the RScript issue is resolved.

**Comment by lloyddewit** [comment]
The third big block of code is fixed in PR #8551 (also new pipe operator and multiline versions).

**Comment by lloyddewit** [comment]
@rdstern I think all the things you asked me to look at above are fixed in PR #8551, please could you test?
Thanks

**Comment by rdstern** [comment]
I just record this here as another sample script with an interesting data set.  Once it is working, it is quite topical, especially as we are likely to be running a workshop in Ukraine.

There is an interesting package with classical historical datasets (histdata).  One of there is from Florence Nightingale about war deaths in Crimea!

I was pleased it is using ggplot2!  It has braces (curly brackets) and weems to have other slight difficulties.  Perhaps it could become an interesting data set for the AIMS course - and generally for us.

```
data(Nightingale)

# For some graphs, it is more convenient to reshape death rates to long format
#  keep only Date and death rates
require(reshape)
Night<- Nightingale[,c(1,8:10)]
melted <- melt(Night, "Date")
names(melted) <- c("Date", "Cause", "Deaths")
melted$Cause <- sub("\\.rate", "", melted$Cause)
melted$Regime <- ordered( rep(c(rep('Before', 12), rep('After', 12)), 3), 
                          levels=c('Before', 'After'))
Night <- melted

# subsets, to facilitate separate plotting
Night1 <- subset(Night, Date < as.Date("1855-04-01"))
Night2 <- subset(Night, Date >= as.Date("1855-04-01"))

# sort according to Deaths in decreasing order, so counts are not obscured [thx: Monique Graf]
Night1 <- Night1[order(Night1$Deaths, decreasing=TRUE),]
Night2 <- Night2[order(Night2$Deaths, decreasing=TRUE),]

# merge the two sorted files
Night <- rbind(Night1, Night2)


require(ggplot2)
# Before plot
cxc1 <- ggplot(Night1, aes(x = factor(Date), y=Deaths, fill = Cause)) +
		# do it as a stacked bar chart first
   geom_bar(width = 1, position="identity", stat="identity", color="black") +
		# set scale so area ~ Deaths	
   scale_y_sqrt() 
		# A coxcomb plot = bar chart + polar coordinates
cxc1 + coord_polar(start=3*pi/2) + 
	ggtitle("Causes of Mortality in the Army in the East") + 
	xlab("")

# After plot
cxc2 <- ggplot(Night2, aes(x = factor(Date), y=Deaths, fill = Cause)) +
   geom_bar(width = 1, position="identity", stat="identity", color="black") +
   scale_y_sqrt()
cxc2 + coord_polar(start=3*pi/2) +
	ggtitle("Causes of Mortality in the Army in the East") + 
	xlab("")

## Not run: 
# do both together, with faceting
cxc <- ggplot(Night, aes(x = factor(Date), y=Deaths, fill = Cause)) +
 geom_bar(width = 1, position="identity", stat="identity", color="black") + 
 scale_y_sqrt() +
 facet_grid(. ~ Regime, scales="free", labeller=label_both)
cxc + coord_polar(start=3*pi/2) +
	ggtitle("Causes of Mortality in the Army in the East") + 
	xlab("")

## End(Not run)

## What if she had made a set of line graphs?

# these plots are best viewed with width ~ 2 * height 
colors <- c("blue", "red", "black")
with(Nightingale, {
	plot(Date, Disease.rate, type="n", cex.lab=1.25, 
		ylab="Annual Death Rate", xlab="Date", xaxt="n",
		main="Causes of Mortality of the British Army in the East");
	# background, to separate before, after
	rect(as.Date("1854/4/1"), -10, as.Date("1855/3/1"), 
		1.02*max(Disease.rate), col=gray(.90), border="transparent");
	text( as.Date("1854/4/1"), .98*max(Disease.rate), "Before Sanitary\nCommission", pos=4);
	text( as.Date("1855/4/1"), .98*max(Disease.rate), "After Sanitary\nCommission", pos=4);
	# plot the data
	points(Date, Disease.rate, type="b", col=colors[1], lwd=3);
	points(Date, Wounds.rate, type="b", col=colors[2], lwd=2);
	points(Date, Other.rate, type="b", col=colors[3], lwd=2)
	}
)
# add custom Date axis and legend
axis.Date(1, at=seq(as.Date("1854/4/1"), as.Date("1856/3/1"), "3 months"), format="%b %Y")
legend(as.Date("1855/10/20"), 700, c("Preventable disease", "Wounds and injuries", "Other"),
	col=colors, fill=colors, title="Cause", cex=1.25)

# Alternatively, show each cause of death as percent of total
Nightingale <- within(Nightingale, {
	Total <- Disease + Wounds + Other
	Disease.pct <- 100*Disease/Total
	Wounds.pct <- 100*Wounds/Total
	Other.pct <- 100*Other/Total
	})

colors <- c("blue", "red", "black")
with(Nightingale, {
	plot(Date, Disease.pct, type="n",  ylim=c(0,100), cex.lab=1.25,
		ylab="Percent deaths", xlab="Date", xaxt="n",
		main="Percentage of Deaths by Cause");
	# background, to separate before, after
	rect(as.Date("1854/4/1"), -10, as.Date("1855/3/1"), 
		1.02*max(Disease.rate), col=gray(.90), border="transparent");
	text( as.Date("1854/4/1"), .98*max(Disease.pct), "Before Sanitary\nCommission", pos=4);
	text( as.Date("1855/4/1"), .98*max(Disease.pct), "After Sanitary\nCommission", pos=4);
	# plot the data
	points(Date, Disease.pct, type="b", col=colors[1], lwd=3);
	points(Date, Wounds.pct, type="b", col=colors[2], lwd=2);
	points(Date, Other.pct, type="b", col=colors[3], lwd=2)
	}
)
# add custom Date axis and legend
axis.Date(1, at=seq(as.Date("1854/4/1"), as.Date("1856/3/1"), "3 months"), format="%b %Y")
legend(as.Date("1854/8/20"), 60, c("Preventable disease", "Wounds and injuries", "Other"),
	col=colors, fill=colors, title="Cause", cex=1.25)

```
